node/core/dispute-coordinator/src/backend.rs

```diff
@@ -66,75 +66,77 @@ pub trait Backend {
 /// state of the overlay.
 pub struct OverlayedBackend<'a, B: 'a> {
 	inner: &'a B,
-
-	// `None` means unchanged.
-	earliest_session: Option<SessionIndex>,
-	// `None` means unchanged.
-	recent_disputes: Option<RecentDisputes>,
-	// `None` means deleted, missing means query inner.
-	candidate_votes: HashMap<(SessionIndex, CandidateHash), Option<CandidateVotes>>,
+	cache: &'a mut OverlayCache,
+	// Reports read operations.
+	metrics: Metrics,
 }
 
 impl<'a, B: 'a + Backend> OverlayedBackend<'a, B> {
-	pub fn new(backend: &'a B) -> Self {
-		Self {
-			inner: backend,
-			earliest_session: None,
-			recent_disputes: None,
-			candidate_votes: HashMap::new(),
-		}
-	}
-
-	/// Returns true if the are no write operations to perform.
-	pub fn is_empty(&self) -> bool {
-		self.earliest_session.is_none() &&
-			self.recent_disputes.is_none() &&
-			self.candidate_votes.is_empty()
+	pub fn new(backend: &'a B, metrics: Metrics, cache: &'a mut OverlayCache) -> Self {
+		Self { inner: backend, cache, metrics }
 	}
 
 	/// Load the earliest session, if any.
 	pub fn load_earliest_session(&self) -> SubsystemResult<Option<SessionIndex>> {
-		if let Some(val) = self.earliest_session {
-			return Ok(Some(val))
+		if let Some(val) = self.cache.get_earliest_session() {
+			return Ok(Some(*val))
 		}
 
+		let _read_timer = self.metrics.time_db_read_operation("earliest_session");
 		self.inner.load_earliest_session()
 	}
 
 	/// Load the recent disputes, if any.
-	pub fn load_recent_disputes(&self) -> SubsystemResult<Option<RecentDisputes>> {
-		if let Some(val) = &self.recent_disputes {
-			return Ok(Some(val.clone()))
+	pub fn load_recent_disputes(&mut self) -> SubsystemResult<Option<RecentDisputes>> {
+		if let Some(disputes) = self.cache.load_recent_disputes() {
+			return Ok(Some(disputes))
 		}
 
-		self.inner.load_recent_disputes()
+		let read_timer = self.metrics.time_db_read_operation("recent_disputes");
+		let disputes = self.inner.load_recent_disputes()?;
+		drop(read_timer);
```

Comment:
For a fair comparision, we should probably include the cache update?

Reply:
yeah, I missed that.

Reply:
Looking at it now - the questions is do we want to include the in-memory update time in the db read operation ? If so, I must measure the entire time spent in this function.